core/src/main/java/com/linecorp/armeria/client/retry/HttpStatusBasedRetryStrategy.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+
+/**
+ * Provides the {@link RetryStrategy} that decides to retry the request again using HTTP statuses.
+ */
+public class HttpStatusBasedRetryStrategy implements RetryStrategy<HttpRequest, HttpResponse> {
+
+    private final Iterable<HttpStatus> retryStatuses;
+
+    /**
+     * Create a new instance that decides to retry the request if the responded status matches one of the
+     * statuses in {@code statusList}.
+     *
+     * @param statusList the list of HTTP status to retry.
+     */
+    protected HttpStatusBasedRetryStrategy(Iterable<HttpStatus> statusList) {
+        this.retryStatuses = statusList;
```

Comment:
- How about using the same name for parameter and member field?
- `requireNonNull()`
- Make sure the specified status list is not empty and does not contain informational statuses.
- Make an immutable copy or create an array.